### ISSUE TITLE
gradient of `einsum` is incorrect for complex numbers

### ISSUE THREAD

**Issue by Krastanov** [tensorflow/tensorflow#37307]
**System information** 
- Attached is a small script reproducing the problem
- OS Platform and Distribution: Ubuntu 18.04 LTS
- TensorFlow installed from pip
- TensorFlow version: v2.1.0-rc2-17-ge5bf8de 2.1.0
- Python version: Python 3.7.5
- Observed both on CPU and GPU

**Describe the current behavior**

The gradient of this `tf.matmul` expression with respect to `p` is computed correctly:

```python
# `h` is an NxN complex128 matrix, `p` is float64 number, and `zero` is a float64 zero
def f(p,h):
    h1 = tf.complex(p,zero) * h
    return tf.abs(tf.reduce_sum(tf.matmul(h,h)))
```

But the value of the following `tf.einsum` expression is the same (and computed correctly), while the gradient with respect to `p` is wrong:

```python
def f(p,h):
    h1 = tf.complex(p,zero) * h
    return tf.abs(tf.reduce_sum(tf.einsum('ab,bc->ac',h,h)))
```

The problem happens only when the matrices are complex.

**Describe the expected behavior**

The two functions should produce the same value (which is working fine), and their gradients with respect to `p` should be the same (which is not happening).

**Standalone code to reproduce the issue** 

https://colab.research.google.com/drive/1FdC-x5Q74NqoLiB6kR7Omtcb5npdrDtx

**Comment by ziofil** [comment]
I'm also experiencing this issue

**Comment by przemek3141** [comment]
I'm also experiencing this issue, and would love to see it resolved:)

**Comment by scarrazza** [comment]
Same issue for me.

**Comment by ziofil** [comment]
@drpngx this issue has been blocking several research projects for weeks now, do you know if anyone is looking into a fix?

**Comment by Krastanov** [comment]
@ziofil, FYI `einsum` calls can be replaced by calls to `matmul`/`tensordot` and `transpose`. It seems you use tensorflow in the same type of research work as me, and while it is a bit more clunky to write the code with `matmul` and `transpose` it ended working fine and unblocked my work (and this bug does not exist there).

Of course, this is no reason not to try to fix this bug.

**Comment by josh146** [comment]
Would another solution be to do the following?
```python
from tensorflow.python.ops.special_math_ops import _einsum_v1 as einsum
```

It seems that TensorFlow 2.1 [introduced a significant refactor of `tf.einsum`](https://github.com/tensorflow/tensorflow/commit/4eba4d514e24b6c1fdae69b40c2b958cc2a1cb42) to use the optimized einsum library, and that might be where the bug was introduced. `_einsum_v1` seems to be the original implementation.

**Comment by josh146** [comment]
For example:
```python
from tensorflow.python.ops.special_math_ops import _einsum_v1

import tensorflow as tf


def f1(h):
    return tf.abs(tf.reduce_sum(tf.matmul(h,h)))

def f2(h):
    return tf.abs(tf.reduce_sum(tf.einsum('ab,bc->ac',h,h)))

def f3(h):
    return tf.abs(tf.reduce_sum(_einsum_v1('ab,bc->ac',h,h)))


A = tf.Variable(tf.random.uniform(shape=[2, 2]))
B = tf.random.uniform(shape=[2, 2])

for f in (f1, f2, f3):
    with tf.GradientTape() as tape:
        C = tf.cast(A, dtype=tf.complex64) + tf.cast(B, dtype=tf.complex64)*1j
        loss = f(C)

    grad = tape.gradient(loss, A)
    print(grad)

```

**Comment by alextp** [comment]
@bloops are you aware of this issue?

**Comment by Randl** [comment]
Seems like a number of additional cases fail for complex numbers, which weren't noticed since gradient tests are performed only for `float64`.

**Comment by bloops** [comment]
I wasn't aware of this. I'll take a look...

**Comment by Randl** [comment]
I've tried adding complex tests, many are failing for complex: https://github.com/Randl/tensorflow/commit/f6865e57c93a1d92df30aee4b3756b2979b9125c

**Comment by ziofil** [comment]
> Seems like a number of additional cases fail for complex numbers, which weren't noticed since gradient tests are performed only for `float64`.

@Randl do you mean for einsum or in general?

**Comment by Randl** [comment]
for einsum. I haven't checked other things, but maybe it worth adding complex number to tests in more systematic way
